{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "6facdae3-1a1d-436e-b19a-a4f03f78541a",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "53ab2861-dcf2-4e68-9937-87e1d93ecc80",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/bw/ynbsjqtd7zg9w387gwggdjr00000gn/T/ipykernel_2749/1985001045.py:1: DtypeWarning: Columns (0) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  gtf = pd.read_csv('/Users/nicolejackson/Documents/Oxford/OBDS_Python_Course/Homo_sapiens.GRCh38.107.gtf',\n"
     ]
    }
   ],
   "source": [
    "gtf = pd.read_csv('/Users/nicolejackson/Documents/Oxford/OBDS_Python_Course/Homo_sapiens.GRCh38.107.gtf',\n",
    "                  sep='\\t',\n",
    "                  header=None,\n",
    "                  comment='#')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "a37f7784-5c9d-4b92-b02e-619b50c4b4eb",
   "metadata": {},
   "outputs": [],
   "source": [
    "gtf.columns=['seqname', 'source', 'feature', 'start', 'end', 'score', 'strand', 'frame', 'attribute']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "f2955d3c-751a-4435-9b8a-59655aa5073d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['ensembl_havana', 'havana', 'ensembl', 'havana_tagene',\n",
       "       'ensembl_havana_tagene', 'mirbase', 'ensembl_tagene', 'insdc'],\n",
       "      dtype=object)"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gtf['source'].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "24c4b981-7110-440e-8e25-65da8141ba5f",
   "metadata": {},
   "outputs": [],
   "source": [
    "ens_df = gtf[gtf['source'].str.contains('ensembl_havana')]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "29582ced-6cc9-474b-9a72-9c4dfdbf6d4a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>attribute</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>1471765</td>\n",
       "      <td>1497848</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000160072\"; gene_version \"20\"; ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>transcript</td>\n",
       "      <td>1471765</td>\n",
       "      <td>1497848</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000160072\"; gene_version \"20\"; ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>exon</td>\n",
       "      <td>1471765</td>\n",
       "      <td>1472089</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000160072\"; gene_version \"20\"; ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>CDS</td>\n",
       "      <td>1471885</td>\n",
       "      <td>1472089</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>gene_id \"ENSG00000160072\"; gene_version \"20\"; ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>1471885</td>\n",
       "      <td>1471887</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>gene_id \"ENSG00000160072\"; gene_version \"20\"; ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3370044</th>\n",
       "      <td>21</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>exon</td>\n",
       "      <td>46445284</td>\n",
       "      <td>46445769</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000160299\"; gene_version \"19\"; ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3370045</th>\n",
       "      <td>21</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>CDS</td>\n",
       "      <td>46445284</td>\n",
       "      <td>46445324</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>2</td>\n",
       "      <td>gene_id \"ENSG00000160299\"; gene_version \"19\"; ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3370046</th>\n",
       "      <td>21</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>stop_codon</td>\n",
       "      <td>46445325</td>\n",
       "      <td>46445327</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>gene_id \"ENSG00000160299\"; gene_version \"19\"; ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3370047</th>\n",
       "      <td>21</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>five_prime_utr</td>\n",
       "      <td>46324156</td>\n",
       "      <td>46324228</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000160299\"; gene_version \"19\"; ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3370048</th>\n",
       "      <td>21</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>three_prime_utr</td>\n",
       "      <td>46445328</td>\n",
       "      <td>46445769</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000160299\"; gene_version \"19\"; ...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>714883 rows Ã— 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        seqname          source          feature     start       end score  \\\n",
       "0             1  ensembl_havana             gene   1471765   1497848     .   \n",
       "1             1  ensembl_havana       transcript   1471765   1497848     .   \n",
       "2             1  ensembl_havana             exon   1471765   1472089     .   \n",
       "3             1  ensembl_havana              CDS   1471885   1472089     .   \n",
       "4             1  ensembl_havana      start_codon   1471885   1471887     .   \n",
       "...         ...             ...              ...       ...       ...   ...   \n",
       "3370044      21  ensembl_havana             exon  46445284  46445769     .   \n",
       "3370045      21  ensembl_havana              CDS  46445284  46445324     .   \n",
       "3370046      21  ensembl_havana       stop_codon  46445325  46445327     .   \n",
       "3370047      21  ensembl_havana   five_prime_utr  46324156  46324228     .   \n",
       "3370048      21  ensembl_havana  three_prime_utr  46445328  46445769     .   \n",
       "\n",
       "        strand frame                                          attribute  \n",
       "0            +     .  gene_id \"ENSG00000160072\"; gene_version \"20\"; ...  \n",
       "1            +     .  gene_id \"ENSG00000160072\"; gene_version \"20\"; ...  \n",
       "2            +     .  gene_id \"ENSG00000160072\"; gene_version \"20\"; ...  \n",
       "3            +     0  gene_id \"ENSG00000160072\"; gene_version \"20\"; ...  \n",
       "4            +     0  gene_id \"ENSG00000160072\"; gene_version \"20\"; ...  \n",
       "...        ...   ...                                                ...  \n",
       "3370044      +     .  gene_id \"ENSG00000160299\"; gene_version \"19\"; ...  \n",
       "3370045      +     2  gene_id \"ENSG00000160299\"; gene_version \"19\"; ...  \n",
       "3370046      +     0  gene_id \"ENSG00000160299\"; gene_version \"19\"; ...  \n",
       "3370047      +     .  gene_id \"ENSG00000160299\"; gene_version \"19\"; ...  \n",
       "3370048      +     .  gene_id \"ENSG00000160299\"; gene_version \"19\"; ...  \n",
       "\n",
       "[714883 rows x 9 columns]"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ens_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "639e6438-fe89-4d1f-93e5-eba4516c85e0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(714883, 9)"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ens_df.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "58e51117-dcb1-4207-be35-0b0a2cc8af5c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>source</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>attribute</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>feature</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>CDS</th>\n",
       "      <td>262675</td>\n",
       "      <td>262675</td>\n",
       "      <td>262675</td>\n",
       "      <td>262675</td>\n",
       "      <td>262675</td>\n",
       "      <td>262675</td>\n",
       "      <td>262675</td>\n",
       "      <td>262675</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Selenocysteine</th>\n",
       "      <td>13</td>\n",
       "      <td>13</td>\n",
       "      <td>13</td>\n",
       "      <td>13</td>\n",
       "      <td>13</td>\n",
       "      <td>13</td>\n",
       "      <td>13</td>\n",
       "      <td>13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>exon</th>\n",
       "      <td>286002</td>\n",
       "      <td>286002</td>\n",
       "      <td>286002</td>\n",
       "      <td>286002</td>\n",
       "      <td>286002</td>\n",
       "      <td>286002</td>\n",
       "      <td>286002</td>\n",
       "      <td>286002</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>five_prime_utr</th>\n",
       "      <td>38189</td>\n",
       "      <td>38189</td>\n",
       "      <td>38189</td>\n",
       "      <td>38189</td>\n",
       "      <td>38189</td>\n",
       "      <td>38189</td>\n",
       "      <td>38189</td>\n",
       "      <td>38189</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene</th>\n",
       "      <td>19290</td>\n",
       "      <td>19290</td>\n",
       "      <td>19290</td>\n",
       "      <td>19290</td>\n",
       "      <td>19290</td>\n",
       "      <td>19290</td>\n",
       "      <td>19290</td>\n",
       "      <td>19290</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>start_codon</th>\n",
       "      <td>25891</td>\n",
       "      <td>25891</td>\n",
       "      <td>25891</td>\n",
       "      <td>25891</td>\n",
       "      <td>25891</td>\n",
       "      <td>25891</td>\n",
       "      <td>25891</td>\n",
       "      <td>25891</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>stop_codon</th>\n",
       "      <td>25908</td>\n",
       "      <td>25908</td>\n",
       "      <td>25908</td>\n",
       "      <td>25908</td>\n",
       "      <td>25908</td>\n",
       "      <td>25908</td>\n",
       "      <td>25908</td>\n",
       "      <td>25908</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>three_prime_utr</th>\n",
       "      <td>30120</td>\n",
       "      <td>30120</td>\n",
       "      <td>30120</td>\n",
       "      <td>30120</td>\n",
       "      <td>30120</td>\n",
       "      <td>30120</td>\n",
       "      <td>30120</td>\n",
       "      <td>30120</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>transcript</th>\n",
       "      <td>26795</td>\n",
       "      <td>26795</td>\n",
       "      <td>26795</td>\n",
       "      <td>26795</td>\n",
       "      <td>26795</td>\n",
       "      <td>26795</td>\n",
       "      <td>26795</td>\n",
       "      <td>26795</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                 seqname  source   start     end   score  strand   frame  \\\n",
       "feature                                                                    \n",
       "CDS               262675  262675  262675  262675  262675  262675  262675   \n",
       "Selenocysteine        13      13      13      13      13      13      13   \n",
       "exon              286002  286002  286002  286002  286002  286002  286002   \n",
       "five_prime_utr     38189   38189   38189   38189   38189   38189   38189   \n",
       "gene               19290   19290   19290   19290   19290   19290   19290   \n",
       "start_codon        25891   25891   25891   25891   25891   25891   25891   \n",
       "stop_codon         25908   25908   25908   25908   25908   25908   25908   \n",
       "three_prime_utr    30120   30120   30120   30120   30120   30120   30120   \n",
       "transcript         26795   26795   26795   26795   26795   26795   26795   \n",
       "\n",
       "                 attribute  \n",
       "feature                     \n",
       "CDS                 262675  \n",
       "Selenocysteine          13  \n",
       "exon                286002  \n",
       "five_prime_utr       38189  \n",
       "gene                 19290  \n",
       "start_codon          25891  \n",
       "stop_codon           25908  \n",
       "three_prime_utr      30120  \n",
       "transcript           26795  "
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "(ens_df.groupby('feature')\n",
    " .count()\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "id": "88c5328c-1962-49c2-b792-cc2419b0e004",
   "metadata": {},
   "outputs": [],
   "source": [
    "genes = ens_df[ens_df[\"feature\"].str.contains(\"gene\")]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "id": "a69baf72-93c7-4caa-8da5-823ebe9cfa06",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>attribute</th>\n",
       "      <th>gene_length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>1471765</td>\n",
       "      <td>1497848</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000160072\"; gene_version \"20\"; ...</td>\n",
       "      <td>26083</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>132</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>3069168</td>\n",
       "      <td>3438621</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000142611\"; gene_version \"17\"; ...</td>\n",
       "      <td>369453</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>361</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>2403964</td>\n",
       "      <td>2413797</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000157911\"; gene_version \"11\"; ...</td>\n",
       "      <td>9833</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>517</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>2228319</td>\n",
       "      <td>2310213</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000157933\"; gene_version \"10\"; ...</td>\n",
       "      <td>81894</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>580</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>10472288</td>\n",
       "      <td>10630758</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000142655\"; gene_version \"13\"; ...</td>\n",
       "      <td>158470</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3365880</th>\n",
       "      <td>21</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>6318378</td>\n",
       "      <td>6360415</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000278903\"; gene_version \"5\"; g...</td>\n",
       "      <td>42037</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3365953</th>\n",
       "      <td>21</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>33915534</td>\n",
       "      <td>33977703</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000237945\"; gene_version \"9\"; g...</td>\n",
       "      <td>62169</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3367972</th>\n",
       "      <td>21</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>9068325</td>\n",
       "      <td>9129773</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000188681\"; gene_version \"13\"; ...</td>\n",
       "      <td>61448</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3368345</th>\n",
       "      <td>21</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>14236206</td>\n",
       "      <td>14362754</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000243064\"; gene_version \"10\"; ...</td>\n",
       "      <td>126548</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3369764</th>\n",
       "      <td>21</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>46324124</td>\n",
       "      <td>46445769</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000160299\"; gene_version \"19\"; ...</td>\n",
       "      <td>121645</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>19290 rows Ã— 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        seqname          source feature     start       end score strand  \\\n",
       "0             1  ensembl_havana    gene   1471765   1497848     .      +   \n",
       "132           1  ensembl_havana    gene   3069168   3438621     .      +   \n",
       "361           1  ensembl_havana    gene   2403964   2413797     .      -   \n",
       "517           1  ensembl_havana    gene   2228319   2310213     .      +   \n",
       "580           1  ensembl_havana    gene  10472288  10630758     .      +   \n",
       "...         ...             ...     ...       ...       ...   ...    ...   \n",
       "3365880      21  ensembl_havana    gene   6318378   6360415     .      +   \n",
       "3365953      21  ensembl_havana    gene  33915534  33977703     .      +   \n",
       "3367972      21  ensembl_havana    gene   9068325   9129773     .      -   \n",
       "3368345      21  ensembl_havana    gene  14236206  14362754     .      +   \n",
       "3369764      21  ensembl_havana    gene  46324124  46445769     .      +   \n",
       "\n",
       "        frame                                          attribute  gene_length  \n",
       "0           .  gene_id \"ENSG00000160072\"; gene_version \"20\"; ...        26083  \n",
       "132         .  gene_id \"ENSG00000142611\"; gene_version \"17\"; ...       369453  \n",
       "361         .  gene_id \"ENSG00000157911\"; gene_version \"11\"; ...         9833  \n",
       "517         .  gene_id \"ENSG00000157933\"; gene_version \"10\"; ...        81894  \n",
       "580         .  gene_id \"ENSG00000142655\"; gene_version \"13\"; ...       158470  \n",
       "...       ...                                                ...          ...  \n",
       "3365880     .  gene_id \"ENSG00000278903\"; gene_version \"5\"; g...        42037  \n",
       "3365953     .  gene_id \"ENSG00000237945\"; gene_version \"9\"; g...        62169  \n",
       "3367972     .  gene_id \"ENSG00000188681\"; gene_version \"13\"; ...        61448  \n",
       "3368345     .  gene_id \"ENSG00000243064\"; gene_version \"10\"; ...       126548  \n",
       "3369764     .  gene_id \"ENSG00000160299\"; gene_version \"19\"; ...       121645  \n",
       "\n",
       "[19290 rows x 10 columns]"
      ]
     },
     "execution_count": 72,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "id": "697eb2ea-9ab3-4947-9e20-7cb3698eb863",
   "metadata": {},
   "outputs": [],
   "source": [
    "ens_df = ens_df.assign(gene_length = ens_df['end'] - ens_df['start'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "id": "330a7879-6363-429c-b8a3-124b1e87aa5a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>attribute</th>\n",
       "      <th>gene_length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>1471765</td>\n",
       "      <td>1497848</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000160072\"; gene_version \"20\"; ...</td>\n",
       "      <td>26083</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>transcript</td>\n",
       "      <td>1471765</td>\n",
       "      <td>1497848</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000160072\"; gene_version \"20\"; ...</td>\n",
       "      <td>26083</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>exon</td>\n",
       "      <td>1471765</td>\n",
       "      <td>1472089</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000160072\"; gene_version \"20\"; ...</td>\n",
       "      <td>324</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>CDS</td>\n",
       "      <td>1471885</td>\n",
       "      <td>1472089</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>gene_id \"ENSG00000160072\"; gene_version \"20\"; ...</td>\n",
       "      <td>204</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>1471885</td>\n",
       "      <td>1471887</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>gene_id \"ENSG00000160072\"; gene_version \"20\"; ...</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3370044</th>\n",
       "      <td>21</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>exon</td>\n",
       "      <td>46445284</td>\n",
       "      <td>46445769</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000160299\"; gene_version \"19\"; ...</td>\n",
       "      <td>485</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3370045</th>\n",
       "      <td>21</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>CDS</td>\n",
       "      <td>46445284</td>\n",
       "      <td>46445324</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>2</td>\n",
       "      <td>gene_id \"ENSG00000160299\"; gene_version \"19\"; ...</td>\n",
       "      <td>40</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3370046</th>\n",
       "      <td>21</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>stop_codon</td>\n",
       "      <td>46445325</td>\n",
       "      <td>46445327</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>gene_id \"ENSG00000160299\"; gene_version \"19\"; ...</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3370047</th>\n",
       "      <td>21</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>five_prime_utr</td>\n",
       "      <td>46324156</td>\n",
       "      <td>46324228</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000160299\"; gene_version \"19\"; ...</td>\n",
       "      <td>72</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3370048</th>\n",
       "      <td>21</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>three_prime_utr</td>\n",
       "      <td>46445328</td>\n",
       "      <td>46445769</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000160299\"; gene_version \"19\"; ...</td>\n",
       "      <td>441</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>714883 rows Ã— 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        seqname          source          feature     start       end score  \\\n",
       "0             1  ensembl_havana             gene   1471765   1497848     .   \n",
       "1             1  ensembl_havana       transcript   1471765   1497848     .   \n",
       "2             1  ensembl_havana             exon   1471765   1472089     .   \n",
       "3             1  ensembl_havana              CDS   1471885   1472089     .   \n",
       "4             1  ensembl_havana      start_codon   1471885   1471887     .   \n",
       "...         ...             ...              ...       ...       ...   ...   \n",
       "3370044      21  ensembl_havana             exon  46445284  46445769     .   \n",
       "3370045      21  ensembl_havana              CDS  46445284  46445324     .   \n",
       "3370046      21  ensembl_havana       stop_codon  46445325  46445327     .   \n",
       "3370047      21  ensembl_havana   five_prime_utr  46324156  46324228     .   \n",
       "3370048      21  ensembl_havana  three_prime_utr  46445328  46445769     .   \n",
       "\n",
       "        strand frame                                          attribute  \\\n",
       "0            +     .  gene_id \"ENSG00000160072\"; gene_version \"20\"; ...   \n",
       "1            +     .  gene_id \"ENSG00000160072\"; gene_version \"20\"; ...   \n",
       "2            +     .  gene_id \"ENSG00000160072\"; gene_version \"20\"; ...   \n",
       "3            +     0  gene_id \"ENSG00000160072\"; gene_version \"20\"; ...   \n",
       "4            +     0  gene_id \"ENSG00000160072\"; gene_version \"20\"; ...   \n",
       "...        ...   ...                                                ...   \n",
       "3370044      +     .  gene_id \"ENSG00000160299\"; gene_version \"19\"; ...   \n",
       "3370045      +     2  gene_id \"ENSG00000160299\"; gene_version \"19\"; ...   \n",
       "3370046      +     0  gene_id \"ENSG00000160299\"; gene_version \"19\"; ...   \n",
       "3370047      +     .  gene_id \"ENSG00000160299\"; gene_version \"19\"; ...   \n",
       "3370048      +     .  gene_id \"ENSG00000160299\"; gene_version \"19\"; ...   \n",
       "\n",
       "         gene_length  \n",
       "0              26083  \n",
       "1              26083  \n",
       "2                324  \n",
       "3                204  \n",
       "4                  2  \n",
       "...              ...  \n",
       "3370044          485  \n",
       "3370045           40  \n",
       "3370046            2  \n",
       "3370047           72  \n",
       "3370048          441  \n",
       "\n",
       "[714883 rows x 10 columns]"
      ]
     },
     "execution_count": 70,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ens_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "id": "4953e143-371c-4921-80d8-9e69d1a0af9d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>gene_length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>count</th>\n",
       "      <td>1.929000e+04</td>\n",
       "      <td>1.929000e+04</td>\n",
       "      <td>1.929000e+04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mean</th>\n",
       "      <td>7.373146e+07</td>\n",
       "      <td>7.379912e+07</td>\n",
       "      <td>6.766304e+04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>std</th>\n",
       "      <td>5.582533e+07</td>\n",
       "      <td>5.583178e+07</td>\n",
       "      <td>1.298385e+05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>min</th>\n",
       "      <td>1.447500e+04</td>\n",
       "      <td>4.687000e+04</td>\n",
       "      <td>6.200000e+01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25%</th>\n",
       "      <td>3.110511e+07</td>\n",
       "      <td>3.114020e+07</td>\n",
       "      <td>9.739250e+03</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50%</th>\n",
       "      <td>5.774250e+07</td>\n",
       "      <td>5.775770e+07</td>\n",
       "      <td>2.736300e+04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75%</th>\n",
       "      <td>1.103538e+08</td>\n",
       "      <td>1.104137e+08</td>\n",
       "      <td>7.086475e+04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>max</th>\n",
       "      <td>2.489062e+08</td>\n",
       "      <td>2.489199e+08</td>\n",
       "      <td>2.473538e+06</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              start           end   gene_length\n",
       "count  1.929000e+04  1.929000e+04  1.929000e+04\n",
       "mean   7.373146e+07  7.379912e+07  6.766304e+04\n",
       "std    5.582533e+07  5.583178e+07  1.298385e+05\n",
       "min    1.447500e+04  4.687000e+04  6.200000e+01\n",
       "25%    3.110511e+07  3.114020e+07  9.739250e+03\n",
       "50%    5.774250e+07  5.775770e+07  2.736300e+04\n",
       "75%    1.103538e+08  1.104137e+08  7.086475e+04\n",
       "max    2.489062e+08  2.489199e+08  2.473538e+06"
      ]
     },
     "execution_count": 68,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genes.describe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "id": "a4e9ea3a-bbdb-438a-8728-141400fdea65",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "tail: genes: No such file or directory\n"
     ]
    }
   ],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "id": "67945a0e-48e9-4ba1-96f3-beb6fc147d11",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>attribute</th>\n",
       "      <th>gene_length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2679474</th>\n",
       "      <td>16</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>5239802</td>\n",
       "      <td>7713340</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000078328\"; gene_version \"23\"; ...</td>\n",
       "      <td>2473538</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1297140</th>\n",
       "      <td>7</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>146116002</td>\n",
       "      <td>148420998</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000174469\"; gene_version \"23\"; ...</td>\n",
       "      <td>2304996</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1733179</th>\n",
       "      <td>9</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>8314246</td>\n",
       "      <td>10613002</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000153707\"; gene_version \"19\"; ...</td>\n",
       "      <td>2298756</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1514121</th>\n",
       "      <td>X</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>31097677</td>\n",
       "      <td>33339609</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000198947\"; gene_version \"18\"; ...</td>\n",
       "      <td>2241932</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1850816</th>\n",
       "      <td>11</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>83455012</td>\n",
       "      <td>85627922</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000150672\"; gene_version \"18\"; ...</td>\n",
       "      <td>2172910</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1601567</th>\n",
       "      <td>8</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>2935353</td>\n",
       "      <td>4994972</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000183117\"; gene_version \"20\"; ...</td>\n",
       "      <td>2059619</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3030998</th>\n",
       "      <td>20</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>13995369</td>\n",
       "      <td>16053197</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000172264\"; gene_version \"18\"; ...</td>\n",
       "      <td>2057828</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1104549</th>\n",
       "      <td>6</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>63719980</td>\n",
       "      <td>65707226</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000188107\"; gene_version \"15\"; ...</td>\n",
       "      <td>1987246</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2041854</th>\n",
       "      <td>10</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>65912457</td>\n",
       "      <td>67763637</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000183230\"; gene_version \"18\"; ...</td>\n",
       "      <td>1851180</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2040306</th>\n",
       "      <td>10</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>53802771</td>\n",
       "      <td>55627942</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000150275\"; gene_version \"20\"; ...</td>\n",
       "      <td>1825171</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        seqname          source feature      start        end score strand  \\\n",
       "2679474      16  ensembl_havana    gene    5239802    7713340     .      +   \n",
       "1297140       7  ensembl_havana    gene  146116002  148420998     .      +   \n",
       "1733179       9  ensembl_havana    gene    8314246   10613002     .      -   \n",
       "1514121       X  ensembl_havana    gene   31097677   33339609     .      -   \n",
       "1850816      11  ensembl_havana    gene   83455012   85627922     .      -   \n",
       "1601567       8  ensembl_havana    gene    2935353    4994972     .      -   \n",
       "3030998      20  ensembl_havana    gene   13995369   16053197     .      +   \n",
       "1104549       6  ensembl_havana    gene   63719980   65707226     .      -   \n",
       "2041854      10  ensembl_havana    gene   65912457   67763637     .      -   \n",
       "2040306      10  ensembl_havana    gene   53802771   55627942     .      -   \n",
       "\n",
       "        frame                                          attribute  gene_length  \n",
       "2679474     .  gene_id \"ENSG00000078328\"; gene_version \"23\"; ...      2473538  \n",
       "1297140     .  gene_id \"ENSG00000174469\"; gene_version \"23\"; ...      2304996  \n",
       "1733179     .  gene_id \"ENSG00000153707\"; gene_version \"19\"; ...      2298756  \n",
       "1514121     .  gene_id \"ENSG00000198947\"; gene_version \"18\"; ...      2241932  \n",
       "1850816     .  gene_id \"ENSG00000150672\"; gene_version \"18\"; ...      2172910  \n",
       "1601567     .  gene_id \"ENSG00000183117\"; gene_version \"20\"; ...      2059619  \n",
       "3030998     .  gene_id \"ENSG00000172264\"; gene_version \"18\"; ...      2057828  \n",
       "1104549     .  gene_id \"ENSG00000188107\"; gene_version \"15\"; ...      1987246  \n",
       "2041854     .  gene_id \"ENSG00000183230\"; gene_version \"18\"; ...      1851180  \n",
       "2040306     .  gene_id \"ENSG00000150275\"; gene_version \"20\"; ...      1825171  "
      ]
     },
     "execution_count": 84,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genes.sort_values(by=['gene_length'], ascending=False).head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "id": "54cac3fc-4050-499f-a957-787abd03609f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>attribute</th>\n",
       "      <th>gene_length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2913291</th>\n",
       "      <td>17</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>43131208</td>\n",
       "      <td>43131270</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000285057\"; gene_version \"1\"; g...</td>\n",
       "      <td>62</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2118533</th>\n",
       "      <td>12</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>93083598</td>\n",
       "      <td>93083675</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000256393\"; gene_version \"1\"; g...</td>\n",
       "      <td>77</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2362811</th>\n",
       "      <td>14</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>22852616</td>\n",
       "      <td>22852728</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000212335\"; gene_version \"1\"; g...</td>\n",
       "      <td>112</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>904441</th>\n",
       "      <td>4</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>90838903</td>\n",
       "      <td>90839037</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000187653\"; gene_version \"11\"; ...</td>\n",
       "      <td>134</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1133423</th>\n",
       "      <td>6</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>73373108</td>\n",
       "      <td>73373263</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000235174\"; gene_version \"1\"; g...</td>\n",
       "      <td>155</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>129959</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>27649419</td>\n",
       "      <td>27649610</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000235912\"; gene_version \"1\"; g...</td>\n",
       "      <td>191</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1847776</th>\n",
       "      <td>11</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>82689559</td>\n",
       "      <td>82689768</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000227097\"; gene_version \"5\"; g...</td>\n",
       "      <td>209</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2938518</th>\n",
       "      <td>18</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>52409042</td>\n",
       "      <td>52409257</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000280098\"; gene_version \"1\"; g...</td>\n",
       "      <td>215</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>675285</th>\n",
       "      <td>3</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>186596976</td>\n",
       "      <td>186597200</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000283390\"; gene_version \"1\"; g...</td>\n",
       "      <td>224</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>988497</th>\n",
       "      <td>5</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>62512246</td>\n",
       "      <td>62512482</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000268942\"; gene_version \"2\"; g...</td>\n",
       "      <td>236</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1967063</th>\n",
       "      <td>10</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>29697935</td>\n",
       "      <td>29698174</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000225976\"; gene_version \"5\"; g...</td>\n",
       "      <td>239</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3255568</th>\n",
       "      <td>Y</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>18992467</td>\n",
       "      <td>18992709</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000185275\"; gene_version \"6\"; g...</td>\n",
       "      <td>242</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3363759</th>\n",
       "      <td>21</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>30718525</td>\n",
       "      <td>30718777</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000231068\"; gene_version \"1\"; g...</td>\n",
       "      <td>252</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2274799</th>\n",
       "      <td>12</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>3211663</td>\n",
       "      <td>3211917</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000227081\"; gene_version \"5\"; g...</td>\n",
       "      <td>254</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1268412</th>\n",
       "      <td>7</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>129410113</td>\n",
       "      <td>129410370</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000238072\"; gene_version \"1\"; g...</td>\n",
       "      <td>257</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>105987</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>152622834</td>\n",
       "      <td>152623103</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000185962\"; gene_version \"1\"; g...</td>\n",
       "      <td>269</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2594158</th>\n",
       "      <td>16</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>26032539</td>\n",
       "      <td>26032811</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000230330\"; gene_version \"1\"; g...</td>\n",
       "      <td>272</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3363751</th>\n",
       "      <td>21</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>30642864</td>\n",
       "      <td>30643136</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000206104\"; gene_version \"2\"; g...</td>\n",
       "      <td>272</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2526556</th>\n",
       "      <td>15</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>29730713</td>\n",
       "      <td>29730985</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000234664\"; gene_version \"1\"; g...</td>\n",
       "      <td>272</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3363727</th>\n",
       "      <td>21</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>30616425</td>\n",
       "      <td>30616699</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000244624\"; gene_version \"3\"; g...</td>\n",
       "      <td>274</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        seqname          source feature      start        end score strand  \\\n",
       "2913291      17  ensembl_havana    gene   43131208   43131270     .      +   \n",
       "2118533      12  ensembl_havana    gene   93083598   93083675     .      -   \n",
       "2362811      14  ensembl_havana    gene   22852616   22852728     .      -   \n",
       "904441        4  ensembl_havana    gene   90838903   90839037     .      -   \n",
       "1133423       6  ensembl_havana    gene   73373108   73373263     .      -   \n",
       "129959        1  ensembl_havana    gene   27649419   27649610     .      +   \n",
       "1847776      11  ensembl_havana    gene   82689559   82689768     .      +   \n",
       "2938518      18  ensembl_havana    gene   52409042   52409257     .      +   \n",
       "675285        3  ensembl_havana    gene  186596976  186597200     .      +   \n",
       "988497        5  ensembl_havana    gene   62512246   62512482     .      -   \n",
       "1967063      10  ensembl_havana    gene   29697935   29698174     .      +   \n",
       "3255568       Y  ensembl_havana    gene   18992467   18992709     .      -   \n",
       "3363759      21  ensembl_havana    gene   30718525   30718777     .      -   \n",
       "2274799      12  ensembl_havana    gene    3211663    3211917     .      -   \n",
       "1268412       7  ensembl_havana    gene  129410113  129410370     .      -   \n",
       "105987        1  ensembl_havana    gene  152622834  152623103     .      -   \n",
       "2594158      16  ensembl_havana    gene   26032539   26032811     .      -   \n",
       "3363751      21  ensembl_havana    gene   30642864   30643136     .      +   \n",
       "2526556      15  ensembl_havana    gene   29730713   29730985     .      +   \n",
       "3363727      21  ensembl_havana    gene   30616425   30616699     .      +   \n",
       "\n",
       "        frame                                          attribute  gene_length  \n",
       "2913291     .  gene_id \"ENSG00000285057\"; gene_version \"1\"; g...           62  \n",
       "2118533     .  gene_id \"ENSG00000256393\"; gene_version \"1\"; g...           77  \n",
       "2362811     .  gene_id \"ENSG00000212335\"; gene_version \"1\"; g...          112  \n",
       "904441      .  gene_id \"ENSG00000187653\"; gene_version \"11\"; ...          134  \n",
       "1133423     .  gene_id \"ENSG00000235174\"; gene_version \"1\"; g...          155  \n",
       "129959      .  gene_id \"ENSG00000235912\"; gene_version \"1\"; g...          191  \n",
       "1847776     .  gene_id \"ENSG00000227097\"; gene_version \"5\"; g...          209  \n",
       "2938518     .  gene_id \"ENSG00000280098\"; gene_version \"1\"; g...          215  \n",
       "675285      .  gene_id \"ENSG00000283390\"; gene_version \"1\"; g...          224  \n",
       "988497      .  gene_id \"ENSG00000268942\"; gene_version \"2\"; g...          236  \n",
       "1967063     .  gene_id \"ENSG00000225976\"; gene_version \"5\"; g...          239  \n",
       "3255568     .  gene_id \"ENSG00000185275\"; gene_version \"6\"; g...          242  \n",
       "3363759     .  gene_id \"ENSG00000231068\"; gene_version \"1\"; g...          252  \n",
       "2274799     .  gene_id \"ENSG00000227081\"; gene_version \"5\"; g...          254  \n",
       "1268412     .  gene_id \"ENSG00000238072\"; gene_version \"1\"; g...          257  \n",
       "105987      .  gene_id \"ENSG00000185962\"; gene_version \"1\"; g...          269  \n",
       "2594158     .  gene_id \"ENSG00000230330\"; gene_version \"1\"; g...          272  \n",
       "3363751     .  gene_id \"ENSG00000206104\"; gene_version \"2\"; g...          272  \n",
       "2526556     .  gene_id \"ENSG00000234664\"; gene_version \"1\"; g...          272  \n",
       "3363727     .  gene_id \"ENSG00000244624\"; gene_version \"3\"; g...          274  "
      ]
     },
     "execution_count": 85,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genes.sort_values(by=['gene_length']).head(20)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 95,
   "id": "03b96ccd-6103-4497-b16e-0504bb4f7365",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>attribute</th>\n",
       "      <th>gene_length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>1471765</td>\n",
       "      <td>1497848</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000160072\"; gene_version \"20\"; ...</td>\n",
       "      <td>26083</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>132</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>3069168</td>\n",
       "      <td>3438621</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000142611\"; gene_version \"17\"; ...</td>\n",
       "      <td>369453</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>361</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>2403964</td>\n",
       "      <td>2413797</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000157911\"; gene_version \"11\"; ...</td>\n",
       "      <td>9833</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>517</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>2228319</td>\n",
       "      <td>2310213</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000157933\"; gene_version \"10\"; ...</td>\n",
       "      <td>81894</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>580</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>10472288</td>\n",
       "      <td>10630758</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000142655\"; gene_version \"13\"; ...</td>\n",
       "      <td>158470</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3365880</th>\n",
       "      <td>21</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>6318378</td>\n",
       "      <td>6360415</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000278903\"; gene_version \"5\"; g...</td>\n",
       "      <td>42037</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3365953</th>\n",
       "      <td>21</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>33915534</td>\n",
       "      <td>33977703</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000237945\"; gene_version \"9\"; g...</td>\n",
       "      <td>62169</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3367972</th>\n",
       "      <td>21</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>9068325</td>\n",
       "      <td>9129773</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000188681\"; gene_version \"13\"; ...</td>\n",
       "      <td>61448</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3368345</th>\n",
       "      <td>21</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>14236206</td>\n",
       "      <td>14362754</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000243064\"; gene_version \"10\"; ...</td>\n",
       "      <td>126548</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3369764</th>\n",
       "      <td>21</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>46324124</td>\n",
       "      <td>46445769</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000160299\"; gene_version \"19\"; ...</td>\n",
       "      <td>121645</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>19290 rows Ã— 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        seqname          source feature     start       end score strand  \\\n",
       "0             1  ensembl_havana    gene   1471765   1497848     .      +   \n",
       "132           1  ensembl_havana    gene   3069168   3438621     .      +   \n",
       "361           1  ensembl_havana    gene   2403964   2413797     .      -   \n",
       "517           1  ensembl_havana    gene   2228319   2310213     .      +   \n",
       "580           1  ensembl_havana    gene  10472288  10630758     .      +   \n",
       "...         ...             ...     ...       ...       ...   ...    ...   \n",
       "3365880      21  ensembl_havana    gene   6318378   6360415     .      +   \n",
       "3365953      21  ensembl_havana    gene  33915534  33977703     .      +   \n",
       "3367972      21  ensembl_havana    gene   9068325   9129773     .      -   \n",
       "3368345      21  ensembl_havana    gene  14236206  14362754     .      +   \n",
       "3369764      21  ensembl_havana    gene  46324124  46445769     .      +   \n",
       "\n",
       "        frame                                          attribute  gene_length  \n",
       "0           .  gene_id \"ENSG00000160072\"; gene_version \"20\"; ...        26083  \n",
       "132         .  gene_id \"ENSG00000142611\"; gene_version \"17\"; ...       369453  \n",
       "361         .  gene_id \"ENSG00000157911\"; gene_version \"11\"; ...         9833  \n",
       "517         .  gene_id \"ENSG00000157933\"; gene_version \"10\"; ...        81894  \n",
       "580         .  gene_id \"ENSG00000142655\"; gene_version \"13\"; ...       158470  \n",
       "...       ...                                                ...          ...  \n",
       "3365880     .  gene_id \"ENSG00000278903\"; gene_version \"5\"; g...        42037  \n",
       "3365953     .  gene_id \"ENSG00000237945\"; gene_version \"9\"; g...        62169  \n",
       "3367972     .  gene_id \"ENSG00000188681\"; gene_version \"13\"; ...        61448  \n",
       "3368345     .  gene_id \"ENSG00000243064\"; gene_version \"10\"; ...       126548  \n",
       "3369764     .  gene_id \"ENSG00000160299\"; gene_version \"19\"; ...       121645  \n",
       "\n",
       "[19290 rows x 10 columns]"
      ]
     },
     "execution_count": 95,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 93,
   "id": "414a683e-61c6-4350-9431-a7ad12a59c3a",
   "metadata": {},
   "outputs": [],
   "source": [
    "genes_2 = genes['attribute'].str.split(';', expand=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 94,
   "id": "22ada159-22e6-40a9-bbe8-2473d44d4577",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>gene_id \"ENSG00000160072\"</td>\n",
       "      <td>gene_version \"20\"</td>\n",
       "      <td>gene_name \"ATAD3B\"</td>\n",
       "      <td>gene_source \"ensembl_havana\"</td>\n",
       "      <td>gene_biotype \"protein_coding\"</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>132</th>\n",
       "      <td>gene_id \"ENSG00000142611\"</td>\n",
       "      <td>gene_version \"17\"</td>\n",
       "      <td>gene_name \"PRDM16\"</td>\n",
       "      <td>gene_source \"ensembl_havana\"</td>\n",
       "      <td>gene_biotype \"protein_coding\"</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>361</th>\n",
       "      <td>gene_id \"ENSG00000157911\"</td>\n",
       "      <td>gene_version \"11\"</td>\n",
       "      <td>gene_name \"PEX10\"</td>\n",
       "      <td>gene_source \"ensembl_havana\"</td>\n",
       "      <td>gene_biotype \"protein_coding\"</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>517</th>\n",
       "      <td>gene_id \"ENSG00000157933\"</td>\n",
       "      <td>gene_version \"10\"</td>\n",
       "      <td>gene_name \"SKI\"</td>\n",
       "      <td>gene_source \"ensembl_havana\"</td>\n",
       "      <td>gene_biotype \"protein_coding\"</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>580</th>\n",
       "      <td>gene_id \"ENSG00000142655\"</td>\n",
       "      <td>gene_version \"13\"</td>\n",
       "      <td>gene_name \"PEX14\"</td>\n",
       "      <td>gene_source \"ensembl_havana\"</td>\n",
       "      <td>gene_biotype \"protein_coding\"</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3365880</th>\n",
       "      <td>gene_id \"ENSG00000278903\"</td>\n",
       "      <td>gene_version \"5\"</td>\n",
       "      <td>gene_source \"ensembl_havana\"</td>\n",
       "      <td>gene_biotype \"lncRNA\"</td>\n",
       "      <td></td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3365953</th>\n",
       "      <td>gene_id \"ENSG00000237945\"</td>\n",
       "      <td>gene_version \"9\"</td>\n",
       "      <td>gene_name \"LINC00649\"</td>\n",
       "      <td>gene_source \"ensembl_havana\"</td>\n",
       "      <td>gene_biotype \"lncRNA\"</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3367972</th>\n",
       "      <td>gene_id \"ENSG00000188681\"</td>\n",
       "      <td>gene_version \"13\"</td>\n",
       "      <td>gene_name \"TEKT4P2\"</td>\n",
       "      <td>gene_source \"ensembl_havana\"</td>\n",
       "      <td>gene_biotype \"transcribed_unprocessed_pseudog...</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3368345</th>\n",
       "      <td>gene_id \"ENSG00000243064\"</td>\n",
       "      <td>gene_version \"10\"</td>\n",
       "      <td>gene_name \"ABCC13\"</td>\n",
       "      <td>gene_source \"ensembl_havana\"</td>\n",
       "      <td>gene_biotype \"transcribed_unprocessed_pseudog...</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3369764</th>\n",
       "      <td>gene_id \"ENSG00000160299\"</td>\n",
       "      <td>gene_version \"19\"</td>\n",
       "      <td>gene_name \"PCNT\"</td>\n",
       "      <td>gene_source \"ensembl_havana\"</td>\n",
       "      <td>gene_biotype \"protein_coding\"</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>19290 rows Ã— 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                 0                   1  \\\n",
       "0        gene_id \"ENSG00000160072\"   gene_version \"20\"   \n",
       "132      gene_id \"ENSG00000142611\"   gene_version \"17\"   \n",
       "361      gene_id \"ENSG00000157911\"   gene_version \"11\"   \n",
       "517      gene_id \"ENSG00000157933\"   gene_version \"10\"   \n",
       "580      gene_id \"ENSG00000142655\"   gene_version \"13\"   \n",
       "...                            ...                 ...   \n",
       "3365880  gene_id \"ENSG00000278903\"    gene_version \"5\"   \n",
       "3365953  gene_id \"ENSG00000237945\"    gene_version \"9\"   \n",
       "3367972  gene_id \"ENSG00000188681\"   gene_version \"13\"   \n",
       "3368345  gene_id \"ENSG00000243064\"   gene_version \"10\"   \n",
       "3369764  gene_id \"ENSG00000160299\"   gene_version \"19\"   \n",
       "\n",
       "                                     2                              3  \\\n",
       "0                   gene_name \"ATAD3B\"   gene_source \"ensembl_havana\"   \n",
       "132                 gene_name \"PRDM16\"   gene_source \"ensembl_havana\"   \n",
       "361                  gene_name \"PEX10\"   gene_source \"ensembl_havana\"   \n",
       "517                    gene_name \"SKI\"   gene_source \"ensembl_havana\"   \n",
       "580                  gene_name \"PEX14\"   gene_source \"ensembl_havana\"   \n",
       "...                                ...                            ...   \n",
       "3365880   gene_source \"ensembl_havana\"          gene_biotype \"lncRNA\"   \n",
       "3365953          gene_name \"LINC00649\"   gene_source \"ensembl_havana\"   \n",
       "3367972            gene_name \"TEKT4P2\"   gene_source \"ensembl_havana\"   \n",
       "3368345             gene_name \"ABCC13\"   gene_source \"ensembl_havana\"   \n",
       "3369764               gene_name \"PCNT\"   gene_source \"ensembl_havana\"   \n",
       "\n",
       "                                                         4     5  \n",
       "0                            gene_biotype \"protein_coding\"        \n",
       "132                          gene_biotype \"protein_coding\"        \n",
       "361                          gene_biotype \"protein_coding\"        \n",
       "517                          gene_biotype \"protein_coding\"        \n",
       "580                          gene_biotype \"protein_coding\"        \n",
       "...                                                    ...   ...  \n",
       "3365880                                                     None  \n",
       "3365953                              gene_biotype \"lncRNA\"        \n",
       "3367972   gene_biotype \"transcribed_unprocessed_pseudog...        \n",
       "3368345   gene_biotype \"transcribed_unprocessed_pseudog...        \n",
       "3369764                      gene_biotype \"protein_coding\"        \n",
       "\n",
       "[19290 rows x 6 columns]"
      ]
     },
     "execution_count": 94,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genes_2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 104,
   "id": "cee27cad-06f9-471a-b34a-eed905001e19",
   "metadata": {},
   "outputs": [],
   "source": [
    "genes_3 = genes.join(genes_2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a3531708-7dec-406b-a01a-b8600a926550",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
